### PR TITLE
[FR] Add ES|QL Custom Library for Rule Support

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issue

Resolves https://github.com/elastic/ia-trade-team/issues/186

## Overview
This is a temporary dev branch for ES|QL exploration and custom library implementation. This PR is meant to stay in draft but serve as delta visualization for ES|QL library implementation and dev.

Related: https://github.com/elastic/ia-trade-team/issues/172

### CODE REVIEW

File: detection_rules/devtools.py [hunk 1]
```diff
@@ -1488,3 +1489,68 @@ def guide_plugin_to_rule(ctx: click.Context, rule_path: Path, save: bool = True)
...
+
+
+@esql_group.command('build-parser')
+@click.pass_context
+def build_parser(antlr_jar: str):
+    """Build the ESQL parser using ANTLR."""
+    # Define paths
+    lexer_file = Path(ESQL_DIR) / 'grammar' / 'EsqlBaseLexer.g4'
+    parser_file = Path(ESQL_DIR) / 'grammar' / 'EsqlBaseParser.g4'
+    antlr_file = Path(get_path('detection_rules')) / 'etc' / 'antlr-4.13.1-complete.jar'
```
Comment (Mikaayenson) [inline detection_rules/devtools.py:1530]: Reminder to delete this.

**Comment by Mikaayenson** [comment]
## Update Nov 27
There is still much to do. Just opening the initial changes with semantic validation and parser generated via antlr4. Some antlr4 files from the initial commit will be removed later.

File: rules/linux/credential_access_credential_dumping_via_unshadow.toml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+[metadata]
```
Comment (Mikaayenson) [inline rules/linux/credential_access_credential_dumping_via_unshadow.toml:1]: ⚠️ DO NOT MERGE WITH THIS RULE ⚠️

File: esql/esql_listener.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        # Determine the type of the literal based on the context type
+        if context_type == 'Comparison' or context_type == 'LogicalIn':
+            if isinstance(ctx, EsqlBaseParser.StringLiteralContext):
+                return 'keyword'  # currently a 'string'
+            elif isinstance(ctx, (EsqlBaseParser.IntegerLiteralContext, EsqlBaseParser.QualifiedIntegerLiteralContext)):
+                return 'integer'
+            elif isinstance(ctx, EsqlBaseParser.DecimalLiteralContext):
+                return 'decimal'
+            elif isinstance(ctx, EsqlBaseParser.BooleanLiteralContext):
+                return 'boolean'
```
Comment (Mikaayenson) [inline esql/esql_listener.py:88]: Need to check these other ecs types

File: rules/integrations/okta/initial_access_success_okta_sso_esql.toml [hunk 4]
```diff
@@ -0,0 +1,92 @@
+[metadata]
```
Comment (Mikaayenson) [inline rules/integrations/okta/initial_access_success_okta_sso_esql.toml:1]: ⚠️ Reminder to self: DELETE THIS FILE ⚠️

File: tests/test_esql_rules.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+            # `process.names` should be `process.name`
+            ('from .ds-logs-endpoint.events.process-default-* [metadata _id, _version, _index] | where process.names like "Microsoft*"',  # noqa: E501
+             pytest.raises(ESQLSemanticError), r"ESQL semantic error: Invalid field: process.names"),
+
+            # Missing `[metadata _id, _version, _index]` without stats
+            ('from .ds-logs-endpoint.events.process-default-* | where process.name like "Microsoft*"',
+             pytest.raises(ESQLSemanticError), r"ESQL semantic error: Missing metadata for ES|QL query with no stats command"),  # noqa: E501
+
+            # returns 0 because count on non-forwarded field process.parent.name
```
Comment (Mikaayenson) [inline tests/test_esql_rules.py:51]: Sample unit tests not yet supported.

**Comment by eric-forte-elastic** [comment]
> This is a temporary dev branch for ES|QL exploration and custom library implementation. This PR is meant to stay in draft but serve as delta visualization for ES|QL library implementation and dev.

Should this PR be put back into draft?

File: esql/utils.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+# from esql.EsqlBaseParser import EsqlBaseParser
+
+
+# def print_tree(parser: EsqlBaseParser, ctx: ParserRuleContext):
+def print_tree(parser, ctx: ParserRuleContext):
+    """Print the parse tree."""
+    print(Trees.toStringTree(ctx, None, parser))
+
+
+# def pretty_print_tree(ctx: EsqlBaseParser.SingleStatementContext, indent: int = 0, is_last: bool = True):
```
Comment (eric-forte-elastic) [inline esql/utils.py:17]: Reminder, I think we can remove these commented out function headers and add the type hinting to the active headers? 
```suggestion

```

**Review by eric-forte-elastic — CHANGES_REQUESTED** [review]

**Pull request closed without merging**